### ISSUE TITLE
Perforce Request Error

### ISSUE THREAD

**Issue by MOST2K2** [groboclown/p4ic4idea#217]
Hello i get this error in Android-Studio with Perforce Plugin.

11:58	Perforce Server Error: Request error: file tmp file operating system write error: Translation of file content failed. (code s:8/g:34/c:0)

Debug:

ient:192.168.0.180:20666@UserCloudSync generated an error. 
com.perforce.p4java.exception.RequestException: file tmp file operating system write error: Translation of file content failed.
	at com.perforce.p4java.impl.mapbased.server.cmd.ResultMapParser.handleErrors(ResultMapParser.java:168)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:562)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:483)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:164)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2297)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:425)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:61)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:124)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:578)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:596)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:237)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2020-05-25 13:10:58,563 [  19736]   WARN - .components.UserErrorComponent - file tmp file operating system write error: Translation of file content failed. 
com.perforce.p4java.exception.RequestException: file tmp file operating system write error: Translation of file content failed.
	at com.perforce.p4java.impl.mapbased.server.cmd.ResultMapParser.handleErrors(ResultMapParser.java:168)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:562)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:483)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:164)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2297)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:425)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:61)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:124)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:578)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:596)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:237)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2020-05-25 13:10:58,567 [  19740]   INFO - n.impl.SimpleConnectionManager - Command execution failed 
net.groboclown.p4.server.api.P4CommandRunner$ServerResultException: file tmp file operating system write error: Translation of file content failed.
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.createServerResultException(P4RequestErrorHandler.java:132)
	at net.groboclown.p4.server.impl.connection.impl.MessageP4RequestErrorHandler.handleException(MessageP4RequestErrorHandler.java:412)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:162)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:578)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:596)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:237)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.perforce.p4java.exception.RequestException: file tmp file operating system write error: Translation of file content failed.
	at com.perforce.p4java.impl.mapbased.server.cmd.ResultMapParser.handleErrors(ResultMapParser.java:168)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:562)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:483)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:164)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2297)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:425)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:61)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:124)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)

Perforce Plugin Verison:
https://github.com/groboclown/p4ic4idea/releases/tag/v0.10.16

Perforce Client:
Rev. P4V/NTX64/2020.1/1946989

Perforce Admin:
Rev. P4Admin/NTX64/2020.1/1946989

Android Studio 4.1 Canary 9:
Build #AI-201.3803.71.41.6466190, built on May 6, 2020
Runtime version: 1.8.0_242-release-1644-b01 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 10 10.0
GC: PS Scavenge, PS MarkSweep
Memory: 4639M
Cores: 12
Registry: ide.new.welcome.screen.force=true, ide.balloon.shadow.size=0
Non-Bundled Plugins: PerforceIC

**Comment by MOST2K2** [comment]
Any news here groboclown? Thanks for investigating!

**Comment by groboclown** [comment]
A little digging later, the error message:

```
Perforce Server Error: Request error: file tmp file operating system write error: Translation of file content failed. (code s:8/g:34/c:0)
```

is exactly that - an error returned by the p4d server itself.  The request was made to the server to get the file contents, but the server encountered this `tmp file operating system write error` and returned that error to the client.

To understand the underlying cause, you'll need to contact your Perforce administrator to investigate the log messages.  There's a chance that the plugin is sending some options that cause the server to do weird things (I've seen some versions of p4d do that), but this particular error message makes me think this could be a out-of-disk-space issue, a virus scanner blocking access, file permissions set wrong, or an issue with the disk itself.

**Comment by MOST2K2** [comment]
I am the perforce admin, too ;)
What logs do you want to see? Its a debian buster linux with enough disk space.
Also i get "your session has expired" every 12 hours although i have setup no password security.
![p4server-info](https://user-images.githubusercontent.com/3817309/84110334-f558bf00-aa24-11ea-97b1-4bf9c1c3dc50.PNG)

**Comment by MOST2K2** [comment]
Here is the log file, its a encoding problem, the string.xml is text here as filetype. I have enabled unicode support at installation of perforce server.

2020-07-17 06:14:32,303 [ 824048]   INFO -                             p4 - cmd (1): p4 info 
2020-07-17 06:14:32,306 [ 824051]   INFO -                             p4 - cmd (2): p4 client -o osusernameCloudSync 
2020-07-17 06:14:32,310 [ 824055]   INFO -                             p4 - cmd (3): p4 where H:/AndroidStudioProjects/App/app/src/main/res/values/strings.xml 
2020-07-17 06:14:32,311 [ 824056]   INFO -                             p4 - cmd (4): p4 print -q //depot/AndroidStudioProjects/App/app/src/main/res/values/strings.xml#9 
2020-07-17 06:14:32,315 [ 824060]   INFO -                         p4.api - p4java error 
java.nio.charset.UnmappableCharacterException: Input length = 1
	at java.nio.charset.CoderResult.throwException(CoderResult.java:282)
	at java.nio.charset.CharsetEncoder.encode(CharsetEncoder.java:816)
	at com.perforce.p4java.CharsetConverter.convert(CharsetConverter.java:100)
	at com.perforce.p4java.CharsetConverter.convert(CharsetConverter.java:186)
	at com.perforce.p4java.impl.mapbased.rpc.func.client.ClientSystemFileCommands.translate(ClientSystemFileCommands.java:671)
	at com.perforce.p4java.impl.mapbased.rpc.func.client.ClientSystemFileCommands.writeText(ClientSystemFileCommands.java:777)
	at com.perforce.p4java.impl.mapbased.rpc.func.client.ClientFunctionDispatcher.dispatch(ClientFunctionDispatcher.java:292)
	at com.perforce.p4java.impl.mapbased.rpc.packet.RpcPacketDispatcher.dispatch(RpcPacketDispatcher.java:160)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:545)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:478)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:166)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2290)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:428)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:70)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:125)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:579)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:597)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:255)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2020-07-17 06:14:32,316 [ 824061]   WARN - nnection.P4RequestErrorHandler - Running an action with the Perforce server client:192.168.0.180:666@osusernameCloudSync generated an error. 
com.perforce.p4java.exception.RequestException: file tmp file operating system write error: Translation of file content failed.
	at com.perforce.p4java.impl.mapbased.server.cmd.ResultMapParser.handleErrors(ResultMapParser.java:191)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:557)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:478)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:166)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2290)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:428)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:70)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:125)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:579)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:597)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:255)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2020-07-17 06:14:32,317 [ 824062]   WARN - .components.UserErrorComponent - file tmp file operating system write error: Translation of file content failed. 
com.perforce.p4java.exception.RequestException: file tmp file operating system write error: Translation of file content failed.
	at com.perforce.p4java.impl.mapbased.server.cmd.ResultMapParser.handleErrors(ResultMapParser.java:191)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:557)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:478)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:166)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2290)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:428)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:70)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:125)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:579)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:597)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:255)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2020-07-17 06:14:32,317 [ 824062]   INFO - n.impl.SimpleConnectionManager - Command execution failed 
net.groboclown.p4.server.api.P4CommandRunner$ServerResultException: file tmp file operating system write error: Translation of file content failed.
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.createServerResultException(P4RequestErrorHandler.java:132)
	at net.groboclown.p4.server.impl.connection.impl.MessageP4RequestErrorHandler.handleException(MessageP4RequestErrorHandler.java:412)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:162)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handle(P4RequestErrorHandler.java:55)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$handleAsync$5(SimpleConnectionManager.java:579)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$startPromise$8(SimpleConnectionManager.java:597)
	at net.groboclown.p4.server.api.async.AsyncAnswer.lambda$background$0(AsyncAnswer.java:49)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:255)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.perforce.p4java.exception.RequestException: file tmp file operating system write error: Translation of file content failed.
	at com.perforce.p4java.impl.mapbased.server.cmd.ResultMapParser.handleErrors(ResultMapParser.java:191)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:557)
	at com.perforce.p4java.impl.mapbased.rpc.NtsServerImpl.execStreamCmd(NtsServerImpl.java:478)
	at com.perforce.p4java.impl.mapbased.server.cmd.BaseDelegator.execStreamCmd(BaseDelegator.java:166)
	at com.perforce.p4java.impl.mapbased.server.cmd.PrintDelegator.getFileContents(PrintDelegator.java:56)
	at com.perforce.p4java.impl.mapbased.server.Server.getFileContents(Server.java:2290)
	at net.groboclown.p4.server.impl.connection.impl.P4CommandUtil.loadContents(P4CommandUtil.java:459)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$null$34(ConnectCommandRunner.java:428)
	at net.groboclown.p4.server.api.commands.file.GetFileContentsQuery.when(GetFileContentsQuery.java:70)
	at net.groboclown.p4.server.impl.connection.ConnectCommandRunner.lambda$getFileContents$37(ConnectCommandRunner.java:401)
	at net.groboclown.p4.server.impl.connection.impl.SimpleConnectionManager.lambda$null$0(SimpleConnectionManager.java:125)
	at net.groboclown.p4.server.impl.connection.P4RequestErrorHandler.handleConnection(P4RequestErrorHandler.java:157)
	... 11 more
2020-07-17 06:14:35,327 [ 827072]   INFO - er.impl.tasks.TempFileWatchDog - Could not delete temporary Perforce file C:\Users\osusername\AppData\Local\Temp\tmp8890610972748147960\p4j1480627477208434666.tmp

**Comment by MOST2K2** [comment]
any news here groboclown?

**Comment by p4paul** [comment]
This looks like a P4Java error or corrupt file; to investigate further we would need:
- P4Java version used in the plugin
- P4D version string from `p4 info` (including OS string)
- P4D Unicode (on/off) and Client OS info
- The 'string.xml' file synced using the command-line and zipped up to preserve the encoding (especially if you use Windows).
- Storage details of the file `p4 fstat -Ol string.xml`

If any of the above is confidential please send it to support@perforce.com.

**Comment by groboclown** [comment]
I've been swamped with real life stuff, but I should have a bit of time at the end of this week.  I know there are a few areas in the code where the file encoding isn't spot-on, so there might be something there where the server expects a utf-8 or similar encoding but the client library is using native encoding.

**Comment by groboclown** [comment]
The interesting line here is:

```
java.nio.charset.UnmappableCharacterException: Input length = 1
```

I looked through the code, and there's plenty of places where file encoding could be improved.  But this issue is on decoding the contents downloaded from the server, and the encoding here comes from the server's description of the encoding.

Does the server's content of the file look okay?  The issue may have been earlier in the command chain with the original submission of the file to the server.  I'm still looking through that side of the code.

**Comment by MOST2K2** [comment]
The server file openend with notepad++ from p4 client shows utf-8 without BOM, looks fine.

**Comment by groboclown** [comment]
So, you're saying that the `p4 print -q //depot/AndroidStudioProjects/App/app/src/main/res/values/strings.xml#9` command generates a utf-8 without BOM file?

I'm adding a bit of logging to the P4Java library, so that it outputs the charset it's converting from.  This should help with the debugging of the code.

On the submit side of things, I inspected the code and found:

1. Performing "Edit" through the explicit request (context menu action) uses the IDE's interpretation of the file's charset.  This seems like the correct behavior.
2. Performing an "edit" implicitly by changing the file contents in an editor (changing the file mode from read-only to writable) does not tell the server the charset to use, so it uses the existing encoding for the file.  I think this is wrong - it means the server expects the encoding of the file it receives to match the stored encoding of the file, and that may not be correct.
3. Performing an Add or Edit or Move via the standard IDE usage (creating a file or other file operations) did not set the encoding, which means the server uses its internal default encoding, and that may not match up with what the IDE uses.

Note that there are 3 encodings here - the client's storage of the file, the Perforce server's record of the current charset, and the server's storage of the file.  On top of this, the IDE has an idea of the local storage encoding, and the environment setup can define a charset (client setup).

Looking through the P4Java library's file upload code (`ClientSendFile.java`), that code loads the contents from disk and sends them to the server.  It looks at the file type known by the server, the explicit charset reported by the server, and whether the server has unicode support enabled.  It (`RpcInputStream.java`) uses this information to read the contents off disk, rather than whatever the IDE or client charset is using.

Based on this, I'm going to change these charset encoding settings to keep the encoding unspecified (`null`) - meaning always use the server's idea of encoding - unless there's a flag set, in which case the IDE encoding will be set to the file on write.

**Comment by MOST2K2** [comment]
No that is not working correctly.
C:\Program Files\Perforce>p4 print -q //depot/AndroidStudioProjects/app/app/src/main/res/values/strings.xml#9
No Translation for parameter 'data' value '3C3F786D6C2076657273696F6E3D22312E302220656E636F64696E673D227574662D38223F3E0A3C7265736F757263657320786D6C6E733A746F6F6C733D22687474703A2F2F736368656D61732E616E64726F69642E636F6D2F746F6F6C73223E0A202020203C737472696E67206E616D653D2277686174736E65772220746....

version 10 is working.
C:\Program Files\Perforce>p4 print -q //depot/AndroidStudioProjects/app/app/src/main/res/values/strings.xml#10
<?xml version="1.0" encoding="utf-8"?>
<resources>
    <string name="whatsnew">
       
I have unicode enabled at the server, should i turn it off?
I have important sources in this workspace of course..

**Comment by groboclown** [comment]
So this is the source of the problem you were experiencing - the P4 plugin attempted to fetch revision 9 (diff views can do that), but the server reports an error, and the plugin bubbles the file encoding error up.

To fix the underlying issue, you may need to do some deep fixing to change its file encoding so the server can return a valid value.

For the IDE issue, there isn't too much to be done here other than making a cleaner user error message.

I'm preparing the next release with the fix for the other encoding problems.

**Comment by MOST2K2** [comment]
ok thanks at the moment its working, i have only changing this file with android studio or notepad++ which work with different encodings.

**Comment by groboclown** [comment]
I'm closing this off, especially since a kind-of fix for it will be in v0.11.0.

**Comment by MOST2K2** [comment]
Ok i will be testing it and give feedback. Thanks
